### PR TITLE
APEXCORE-738 add SSL configuration related documentation to reflect APEXCORE-712

### PR DESCRIPTION
**Pull request by sanjaypujare**
@vrozov @PramodSSImmaneni @tweise pls review and merge as appropriate

### CODE REVIEW

File: docs/security.md [hunk 1]
```diff
@@ -1,15 +1,17 @@
 Security
 ==========
 
-Applications built on Apex run as native YARN applications on Hadoop. The security framework and apparatus in Hadoop apply to the applications. The default security mechanism in Hadoop is Kerberos.
+Applications built on Apex run as native YARN applications on Hadoop. The security framework and apparatus in Hadoop apply to the applications. Both authentication and SSL aspects of the security framework are covered here.
 
-Kerberos Authentication
------------------------
+#Authentication
```
Comment (vrozov) [inline docs/security.md]: Add space after #

File: docs/security.md [hunk 2]
```diff
@@ -1,15 +1,17 @@
 Security
 ==========
 
-Applications built on Apex run as native YARN applications on Hadoop. The security framework and apparatus in Hadoop apply to the applications. The default security mechanism in Hadoop is Kerberos.
+Applications built on Apex run as native YARN applications on Hadoop. The security framework and apparatus in Hadoop apply to the applications. Both authentication and SSL aspects of the security framework are covered here.
```
Comment (pramodin) [inline docs/security.md:5]: SSL as we are supporting currently is for stram web services isn't it. For example, this is not the SSL configuration for RPC or when stram is a web services client. I would consider moving it to web services security.

File: docs/security.md [hunk 3]
```diff
@@ -281,7 +283,58 @@ How does the client get the web service token in the first place? The client wil
...
+
+#### Approach 1: Using Default Hadoop-YARN SSL
+If the default Hadoop-YARN SSL material is uniformly available in the local file system of each node in your Hadoop cluster, STRAM can use it provided the STRAM process has access to the files. Use your Hadoop/YARN configuration procedures to enable SSL in YARN to use this approach, so all YARN applications including Apex applications will be SSL-enabled. However if the SSL material is not present, or not accessible to the STRAM process (e.g. because STRAM does not run as `root`), then you will have to use one of the following 2 approaches.
+
+#### Approach 2: Pre-installing SSL Files on Hadoop Cluster Nodes
+With this approach, your own SSL files are pre-installed on each node in the Hadoop cluster, so the STRAM can access them regardless of the node it runs on. Each node will have 2 files: `ssl-server.xml` is the master SSL configuration file and the SSL keystore file (a JKS file) whose location is specified in `ssl-server.xml`. Decide on a location for each of these files and follow the steps below.
```
Comment (amberarrow) [inline docs/security.md]: whose location is specified => whose name and location are specified

File: docs/security.md [hunk 4]
```diff
@@ -281,7 +283,58 @@ How does the client get the web service token in the first place? The client wil
...
+
+- Create an SSL keystore as described [here](http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html). Assume this file is named myapex-keystore.jks and will reside at `/opt/apex/keystore/` on each Hadoop node and the keystore password is `storepass1` and the keystore key-password is `keypass2`.
```
Comment (amberarrow) [inline docs/security.md]: myapex-keystore.jks => surround with backquotes

File: docs/security.md [hunk 5]
```diff
@@ -281,7 +283,58 @@ How does the client get the web service token in the first place? The client wil
...
+    <name>ssl.server.keystore.password</name>
+    <value>storepass1</value>
+    <description></description>
+  </property>
+</configuration>
+```
+
+Let's assume this file will reside on each node as `/opt/apex/sslConfig/my-apex-ssl-server.xml` .
+
+- Copy the keystore file and SSL configuration file on each Hadoop node to their designated locations and make sure these files are readable by the Apex application using them.
```
Comment (amberarrow) [inline docs/security.md]: these files => they

File: docs/security.md [hunk 6]
```diff
@@ -281,7 +283,58 @@ How does the client get the web service token in the first place? The client wil
...
+```
+launch -Ddt.attr.SSL_CONFIG="{\"configPath\":\"/opt/apex/sslConfig/my-apex-ssl-server.xml\"}"  <apa file>
+```
+
+- Alternatively you can use configuration files to supply the value of `dt.attr.SSL-CONFIG` attribute as described [here](http://apex.apache.org/docs/apex/application_packages/#application-packages)
+
+#### Approach 3: Using Apex CLI to Copy the Keystore
+If you cannot pre-install the SSL files on all your cluster nodes, you can use the SSL-CONFIG attribute in such a way that the Apex CLI copies the keystore file from the client node to the server node and also passes on the keystore password values to the STRAM. The keystore file needs to be accessible on the Apex CLI machine so the Apex CLI can copy it. Let's assume this location is `/opt/apexCli/ssl/myapex-keystore.jks`. Use the Apex CLI launch command as follows:
+```
+launch -Ddt.attr.SSL_CONFIG="{\"keyStorePath\":\"/opt/apexCli/ssl/myapex-keystore.jks\", \"keyStorePassword\":\"storepass1\",\"keyStoreKeyPassword\":\"keypass2\"}"  <apa file>
```
Comment (amberarrow) [inline docs/security.md]: space after second comma {or omit space before first}

Comment (vrozov) [reply docs/security.md]: space needs to be escaped, so no space after the second comma

File: docs/security.md (hunk 2 above)
Comment (sanjaypujare) [reply docs/security.md:5]: @PramodSSImmaneni you are right but my point is this page now covers not only (Kerberos) authentication but also SSL as a security aspect. Are you suggesting changing the wording to include "Web services SSL" or something to that effect?

**Comment by sanjaypujare** [comment]
comments incorporated, pls review again

**Comment by amberarrow** [comment]
Looks good.

**Comment by sanjaypujare** [comment]
@amberarrow can you merge it (assuming others are okay)?

**Comment by amberarrow** [comment]
I'll merge in a day or two.

**Comment by sanjaypujare** [comment]
@amberarrow can you merge if everything looks okay?

**Comment by vrozov** [comment]
@sanjaypujare Please see http://apex.apache.org/community.html#contributing. The commits must be rebased and squashed before the merge is requested.

**Comment by sanjaypujare** [comment]
@vrozov  squashed and rebased to master

File: docs/security.md [hunk 7]
```diff
@@ -281,7 +283,58 @@ How does the client get the web service token in the first place? The client wil
...
+</configuration>
+```
+
+Let's assume this file will reside on each node as `/opt/apex/sslConfig/my-apex-ssl-server.xml` .
+
+- Copy the keystore file and the SSL configuration file on each Hadoop node to their designated locations and make sure they are readable by your Apex application.
+
+- Using the Apex CLI, pass the SSL configuration file location to your Apex application using the SSL_CONFIG attribute as follows:
+```
+launch -Ddt.attr.SSL_CONFIG="{\"configPath\":\"/opt/apex/sslConfig/my-apex-ssl-server.xml\"}"  <apa file>
```
Comment (tweise) [inline docs/security.md:326]: @sanjaypujare replace dt. with apex. - you can do that as part of #552

File: docs/security.md [hunk 8]
```diff
@@ -281,7 +283,58 @@ How does the client get the web service token in the first place? The client wil
...
+```
+launch -Ddt.attr.SSL_CONFIG="{\"keyStorePath\":\"/opt/apexCli/ssl/myapex-keystore.jks\",\"keyStorePassword\":\"storepass1\",\"keyStoreKeyPassword\":\"keypass2\"}"  <apa file>
+```
+
+Apec CLI will copy the keystore file `/opt/apexCli/ssl/myapex-keystore.jks` to the destination STRAM node and also pass on the keystore password values to the STRAM. As mentioned above, you can also use configuration files to supply the value of `dt.attr.SSL_CONFIG`.
 
-We looked at the different security configuration options that are available in Apex, saw the different security requirements for distributed applications in a secure Hadoop environment in detail and looked at how the various security mechanisms in Apex solves this.
+#### Dependencies
+The use of the attribute `dt.attr.SSL_CONFIG` described in the last 2 approaches is dependent on enhancements made in YARN (https://issues.apache.org/jira/browse/YARN-6457) and Apex (https://issues.apache.org/jira/browse/APEXCORE-712) so make sure your Hadoop and Apex versions have those enhancements included.
```
Comment (tweise) [inline docs/security.md:340]: Mention of Apex enhancement included isn't really necessary - docs are versioned and this will be released with 3.7 only.

**Pull request merged**